packages/compiler/core/checker.ts

```diff
@@ -2440,12 +2470,12 @@ export function createChecker(program: Program): Checker {
 
     const overriddenProp = getOverriddenProperty(newProp);
     if (overriddenProp) {
-      const [isAssignable, _] = isTypeAssignableTo(newProp.type, overriddenProp.type, newProp);
+      const isAssignable = isPropertyAssignableTo(newProp, overriddenProp, newProp);
       const parentScalar = overriddenProp.type.kind === "Scalar";
       const parentType = getTypeName(overriddenProp.type);
       const newPropType = getTypeName(newProp.type);
 
-      if (!parentScalar) {
+      if (!parentScalar && !isDiscriminator(newProp)) {
```

Comment:
I'm concerned about a decorator participating in overridability. Maybe instead we should loosen it from "literals can override scalars" to "literals can override literals and scalars" instead. I forget why we limited overriding in the first place, but I guess there were some unsolved design questions for the general case of allowing any property override where new property type is subtype of base property type? I think that is the ultimate place we intend to land. Not bringing decorators into the equation between now and then feels better. cc @bterlson

Reply:
Hmm, I guess literal overriding literal of different value would break the general subtyping rule too, and why you needed that special isPropertyAssignableTo helper. That feels even more concerning. The definition of Salmon.kind looks wrong. It claims that anything that is a Salmon has kind "salmon". This seems even more troubling to be honest.

Reply:
This would be a more valid way to express a hierarchy IMHO:


```
@discriminator("kind")
model Fish {
  name: string;
  kind: string;
}

model Salmon extends Fish {
  kind: "salmon" | "smartSalmon";
}

model PlainSalmon extends Salmon {
 kind: "salmon";
}

model SmartSalmon extends Salmon {
  kind: "smartSalmon";
}
```

Reply:
Or Salmon could not override kind and leave that to the leaf nodes.